### ISSUE TITLE
Error while using "&" и "|" at the same time

### ISSUE THREAD

**Issue by s-belichenko** [squizlabs/PHP_CodeSniffer#2806]
Example:

```
/**
     * Blah.
     *
     * @param string|int|null                    $jsonRpcId          Blah.
     * @param RequestIdGenerator&MockObject|null $requestIdGenerator Blah.
     * @param string                             $exceptedId         Blah.
     *
     * @dataProvider correctConstructProvider
     *
     * @throws \Throwable
     */
```

Errors:

> phpcs: Expected 4 spaces after parameter type; 20 found
> 
> phpcs: Expected 7 spaces after parameter name; 10 found
> 
> phpcs: Doc comment for parameter &MockObject|null does not match actual variable name $requestIdGenerator
> 
> phpcs: Expected 1 spaces after parameter type; 0 found
> 
> phpcs: Parameter comment must start with a capital letter
> 
> phpcs: Expected 13 spaces after parameter type; 29 found
> 
> phpcs: Expected 6 spaces after parameter name; 9 found

**Comment by VincentLanglet** [comment]
@s-belichenko I don't know if it will help but you have to know your syntax is not correct.
It should be `(RequestIdGenerator&MockObject)|null`.

> When union and intersection types are in the same expression they must be grouped with parenthesis (demonstrated above).

See https://wiki.php.net/rfc/intersection_types

**Comment by s-belichenko** [comment]
@VincentLanglet thx for your answer, man, but you write about lang rules while I write about annotations (phpDoc). I tried your variant and it doesn’t work in PhpStorm and in Code_Sniffer.

However, I admit that I could be wrong and you can provide full text correctly phpDoc annotation.

**Comment by VincentLanglet** [comment]
> @VincentLanglet thx for your answer, man, but you write about lang rules while I write about annotations (phpDoc).

There is no difference between the lang rule and the annotation rule.
`(RequestIdGenerator&MockObject)|null` is the good syntax.

You can verify with psalm or phpstan, they accept the `(RequestIdGenerator&MockObject)|null` but not the syntax without parenthesis.

> I tried your variant and it doesn’t work in PhpStorm 

That's another issue. Phpstorm doesn't support a lot of correct annotation.
`(int|string)[]` is not supported.
`array<int>` too.
array{key: value} too.

They does support `A&B` magically (but consider this like `A|B`), but does not support parenthesis.

Request was made to Phpstorm to support the new annotations.

> and in Code_Sniffer.

It should have been fixed in 3.5.1.
https://github.com/squizlabs/PHP_CodeSniffer/commit/f797a35dafabd9dddab3029e47f550ba7c615280

What Sniff do you use ? What version do you use ?

**Comment by geerteltink** [comment]
@VincentLanglet I just tested this: `(RequestIdGenerator&MockObject)|null` and `RequestIdGenerator&MockObject|null` give the same result. phpcs corrects it wrong.

**Comment by VincentLanglet** [comment]
> @VincentLanglet I just tested this: `(RequestIdGenerator&MockObject)|null` and `RequestIdGenerator&MockObject|null` give the same result. phpcs corrects it wrong.

What sniff is reporting an error ? What version of phpcs do you use.

This commit was made for this... https://github.com/squizlabs/PHP_CodeSniffer/commit/f797a35dafabd9dddab3029e47f550ba7c615280#diff-c5c153487a24f533f5a67c565e703896R375 was

**Comment by geerteltink** [comment]
Using 3.5.4

Looks like another sniff is responsible for this.

Test class:
```php
<?php

namespace Test;

class Test
{
     /**
      * @param ContainerInterface&ObjectProphecy $container
      * @param mixed              $service
      */
    protected function injectServiceInContainer(ObjectProphecy $container, string $serviceName, $service): void
    {
        $container->has($serviceName)->willReturn(true);
        $container->get($serviceName)->willReturn($service);
    }
}
```

Report:

```text
--------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------------------------
 9 | ERROR | [x] Expected 1 spaces after parameter type; 0 found
   |       |     (Squiz.Commenting.FunctionComment.SpacingAfterParamType)
--------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------------------------------------------------
```

Using these give all the same result:

```
@param ContainerInterface&ObjectProphecy $container
@param ContainerInterface&ObjectProphecy|null $container
@param (ContainerInterface&ObjectProphecy)|null $container
```

**Comment by VincentLanglet** [comment]
@geerteltink Thanks for the described issue.

This is comming from https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php#L283

We need the same change than https://github.com/squizlabs/PHP_CodeSniffer/commit/f797a35dafabd9dddab3029e47f550ba7c615280#diff-c5c153487a24f533f5a67c565e703896R375

Do you want to create the PR ?

I recommend 
- To create a constant or protected method in the src/Standards/PEAR/Sniffs/Commenting/FunctionCommentSniff.php
- To use this constant/method in the src/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php (which is extending the PEAR one)

This way this won't happen again if we update the regex of the PEAR Sniff.

**Comment by rodrigoprimo** [comment]
The initial report about `PEAR.Commenting.FunctionComment` was fixed by https://github.com/squizlabs/PHP_CodeSniffer/commit/f797a35dafabd9dddab3029e47f550ba7c615280.

I believe the follow-up report by @geerteltink about `Squiz.Commenting.FunctionComment` was fixed by https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/944. This is the same issue reported in #2887, which was closed as fixed.

I suggest closing this issue.
